### ISSUE TITLE
Request: Semantic highlighting

### ISSUE THREAD

**Issue by MrSet** [cpitclaudel/company-coq#207]
I have a background writing Idris, but recently transferred to Coq. In both Idris and Agda they have semantic highlighting which I have come to miss.
@spitters and I are both wondering if there are plans/works on extending company-coq with such features?

**Comment by cpitclaudel** [comment]
Not at the moment. Most of the syntax highlighting is done by Proof-General, and it's based on regular expressions, not on proper parsing.

One issue is that semantic highlighting is only well-defined for complete programs, and most of the time (in a non-structured editor) the contents of the current file do not form a complete program.

Last time I used Agda, the Emacs mode worked around this issue by only syntax-highlighting well-formed code.  Has this improved since? Does Idris do better?

**Comment by spitters** [comment]
Not sure whether this is related, but it gives some background on the
idris editor and related work:
https://cattheory.com/extensibleTypeDirectedEditing.pdf

**Comment by MrSet** [comment]
For Agda i dont know, but Idris is still relying on the code to be well-formed to do semantic highlighting. 

Idris do however have a very well developed hole-mechanism (a mechanism for making a introducing a placeholder for non-finished definitions, which the type-checker treats as a wildcard). For details on "holes" see Brady's original paper on Idris: [https://eb.host.cs.st-andrews.ac.uk/drafts/impldtp.pdf](url).

The intended workflow in Idris (as least the workflow advocated by Brady) is to write functions incrementally using these "holes" for non-finished parts, and let the remaining implementation be guided by the types.  
Therefore is the program being written, nearly at all times well-formed and the semantic-highlighting can thus guide you.

**Comment by cpitclaudel** [comment]
Right, that's what I meant by "structured editor" above.  We don't really have this for Coq, unfortunately (or fortunately? I'm not a huge fan of that experience ^^)

**Comment by spitters** [comment]
How about this?
https://stackoverflow.com/questions/41837820/agda-like-programming-in-coq-proof-general

or even the new Equations plug in ?

**Comment by cpitclaudel** [comment]
Hmm, maybe I wasn't clear.  I'm not talking about a limitation of Coq, I'm talking about the way PG works.
Concretely, when you write lambda, you tend to type `( f u n = > …some stuff here… )`.  For most of that time, your file is not in a consistent state.

A structured editor would force you to always have matched braces, and so on.

In other words, with the current PG, the question is: how would we highlight `specialize (fun x (y: na`?

**Comment by spitters** [comment]
I seem to recall that lean just considers this a spelling error (using
flyspeck).
https://github.com/leanprover/lean-mode

**Comment by cpitclaudel** [comment]
> I seem to recall that lean just considers this a spelling error

Sure, but it still highlights the code.  IOW, it does the same thing as PG, doesn't it?

> flyspeck

:) I think you mean Flycheck, but I actually find the name flyspeck a lot cuter ^^

**Comment by simongregersen** [comment]
Wouldn't it be possible not to do semantic highlighting of incomplete terms (like `specialize (fun x (y: na`) but only of terms that are 'loaded' by Coq and therefore complete and type checkable?

**Comment by cpitclaudel** [comment]
> I suppose that the straightforward way to go would be to do  semantic highlighting of code that is actually loaded by the Coq  process

Indeed, that would be completely feasible.
